File: backend/utils/audit_logger.py
from database import get_db
from models import AuditLogCreate, ActionType, ResourceType
import logging
from datetime import datetime
from typing import Optional, Dict, Any

logger = logging.getLogger(__name__)

class AuditLogger:
    """Utility class for logging audit events"""
    
    @staticmethod
    async def log(
        action_type: ActionType,
        resource_type: ResourceType,
        user_id: Optional[str] = None,
        user_email: Optional[str] = None,
        resource_id: Optional[str] = None,
        resource_name: Optional[str] = None,
        details: Optional[Dict[str, Any]] = None,
        ip_address: Optional[str] = None,
        user_agent: Optional[str] = None,
        school_id: Optional[str] = None
    ) -> bool:
        """Log an audit event to the database"""
        try:
            db = get_db()
            
            audit_data = {
                "user_id": user_id,
                "user_email": user_email,
                "action_type": action_type.value if isinstance(action_type, ActionType) else action_type,
                "resource_type": resource_type.value if isinstance(resource_type, ResourceType) else resource_type,
                "resource_id": resource_id,
                "resource_name": resource_name,
                "details": details,
                "ip_address": ip_address,
                "user_agent": user_agent,
                "school_id": school_id,
                "created_at": datetime.utcnow().isoformat()
            }
            
            result = db.table('audit_logs').insert(audit_data).execute()
            
            if result.data:
                logger.info(f"Audit log created: {action_type.value} on {resource_type.value}")
                return True
            else:
                logger.error(f"Failed to create audit log: {result}")
                return False
                
        except Exception as e:
            logger.error(f"Error creating audit log: {str(e)}")
            return False
    
    @staticmethod
    async def log_pdf_view(
        user_id: str,
        user_email: str,
        pdf_path: str,
        exam_subject_id: str,
        ip_address: Optional[str] = None,
        school_id: Optional[str] = None
    ):
        """Convenience method for logging PDF views"""
        await AuditLogger.log(
            action_type=ActionType.VIEW,
            resource_type=ResourceType.PDF,
            user_id=user_id,
            user_email=user_email,
            resource_id=exam_subject_id,
            resource_name=pdf_path,
            ip_address=ip_address,
            school_id=school_id,
            details={"file_path": pdf_path}
        )
    
    @staticmethod
    async def log_pdf_download(
        user_id: str,
        user_email: str,
        pdf_path: str,
        exam_subject_id: str,
        ip_address: Optional[str] = None,
        school_id: Optional[str] = None
    ):
        """Convenience method for logging PDF downloads"""
        await AuditLogger.log(
            action_type=ActionType.DOWNLOAD,
            resource_type=ResourceType.PDF,
            user_id=user_id,
            user_email=user_email,
            resource_id=exam_subject_id,
            resource_name=pdf_path,
            ip_address=ip_address,
            school_id=school_id,
            details={"file_path": pdf_path}
        )

File: frontend/src/utils/auditLogger.ts
import apiClient from '@/lib/apiClient';
import { supabase } from '@/integrations/supabase/client';

export type ActionType = 
  | 'login' 
  | 'logout' 
  | 'upload' 
  | 'view' 
  | 'download' 
  | 'delete' 
  | 'create' 
  | 'update'
  | 'export';

export type ResourceType = 
  | 'pdf' 
  | 'student' 
  | 'exam' 
  | 'score' 
  | 'teacher' 
  | 'report' 
  | 'settings';

export interface AuditLogData {
  action_type: ActionType;
  resource_type: ResourceType;
  resource_id?: string;
  resource_name?: string;
  details?: Record<string, any>;
}

class AuditLoggerService {
  /**
   * Log an audit event
   */
  async log(data: AuditLogData): Promise<void> {
    try {
      const { data: { user } } = await supabase.auth.getUser();
      
      await apiClient.post('/audit/log', {
        user_id: user?.id,
        user_email: user?.email,
        school_id: user?.user_metadata?.school_id,
        ...data,
      });
    } catch (error) {
      console.error('Failed to log audit event:', error);
      // Don't throw - audit logging failure shouldn't break the app
    }
  }

  /**
   * Log PDF view
   */
  async logPdfView(examSubjectId: string, pdfPath: string): Promise<void> {
    await this.log({
      action_type: 'view',
      resource_type: 'pdf',
      resource_id: examSubjectId,
      resource_name: pdfPath,
      details: { file_path: pdfPath },
    });
  }

  /**
   * Log PDF download
   */
  async logPdfDownload(examSubjectId: string, pdfPath: string): Promise<void> {
    await this.log({
      action_type: 'download',
      resource_type: 'pdf',
      resource_id: examSubjectId,
      resource_name: pdfPath,
      details: { file_path: pdfPath },
    });
  }

  /**
   * Log user login
   */
  async logLogin(): Promise<void> {
    await this.log({
      action_type: 'login',
      resource_type: 'teacher',
    });
  }

  /**
   * Log user logout
   */
  async logLogout(): Promise<void> {
    await this.log({
      action_type: 'logout',
      resource_type: 'teacher',
    });
  }

  /**
   * Log student creation
   */
  async logStudentCreate(studentId: string, studentName: string): Promise<void> {
    await this.log({
      action_type: 'create',
      resource_type: 'student',
      resource_id: studentId,
      resource_name: studentName,
    });
  }

  /**
   * Log exam creation
   */
  async logExamCreate(examId: string, examName: string): Promise<void> {
    await this.log({
      action_type: 'create',
      resource_type: 'exam',
      resource_id: examId,
      resource_name: examName,
    });
  }

  /**
   * Log score entry
   */
  async logScoreEntry(scoreId: string, studentName: string, examName: string): Promise<void> {
    await this.log({
      action_type: 'create',
      resource_type: 'score',
      resource_id: scoreId,
      details: { student_name: studentName, exam_name: examName },
    });
  }

  /**
   * Log report export
   */
  async logReportExport(reportType: string, format: string): Promise<void> {
    await this.log({
      action_type: 'export',
      resource_type: 'report',
      resource_name: reportType,
      details: { format },
    });
  }
}

// Export singleton instance
export const auditLogger = new AuditLoggerService();
